File: nfc_manager_android/android/src/main/kotlin/dev/huynh/nfc_manager_android/models/HostCardEmulationStatus.kt
package dev.huynh.nfc_manager_android.models

/**
 * Represents the status of the host card emulation (HCE) operation.
 */
enum class HostCardEmulationStatus {
    /**
     * Indicates that the HCE is ready for operations.
     */
    READY,

    /**
     * Indicates that the command is invalid because it must be at least 4 bytes
     * long.
     */
    INVALID_COMMAND,

    /**
     * Indicates that the length of the command data is invalid.
     */
    INVALID_LC_LENGTH,

    /**
     * Indicates that the AID (Application Identifier) has been selected.
     */
    AID_SELECTED,

    /**
     * Indicates that the AID is invalid.
     */
    INVALID_AID,

    /**
     * Indicates that the PIN has been verified.
     */
    PIN_VERIFIED,

    /**
     * Indicates that the PIN is incorrect.
     */
    WRONG_PIN,

    /**
     * Indicates that the function is not supported.
     */
    FUNCTION_NOT_SUPPORTED,

    /**
     * Indicates that the tag is disconnected.
     */
    TAG_DISCONNECTED,
}

File: nfc_manager_platform_interface/lib/src/models/host_card_emulation_status.dart
/// Represents the status of the host card emulation (HCE) operation.
enum HostCardEmulationStatus {
  /// Indicates that the HCE is ready for operations.
  ready,

  /// Indicates that the command is invalid because it must be at least 4 bytes
  /// long.
  invalidCommand,

  /// Indicates that the length of the command data is invalid.
  invalidLcLength,

  /// Indicates that the AID (Application Identifier) has been selected.
  aidSelected,

  /// Indicates that the AID is invalid.
  invalidAid,

  /// Indicates that the PIN has been verified.
  pinVerified,

  /// Indicates that the PIN is incorrect.
  wrongPin,

  /// Indicates that the function is not supported.
  functionNotSupported,

  /// Indicates that the tag is disconnected.
  tagDisconnected;

  /// Constructs a [HostCardEmulationStatus] from a string representation.
  factory HostCardEmulationStatus.fromString(String status) => switch (status) {
    'READY' => HostCardEmulationStatus.ready,
    'INVALID_COMMAND' => HostCardEmulationStatus.invalidCommand,
    'INVALID_LC_LENGTH' => HostCardEmulationStatus.invalidLcLength,
    'AID_SELECTED' => HostCardEmulationStatus.aidSelected,
    'INVALID_AID' => HostCardEmulationStatus.invalidAid,
    'PIN_VERIFIED' => HostCardEmulationStatus.pinVerified,
    'WRONG_PIN' => HostCardEmulationStatus.wrongPin,
    'FUNCTION_NOT_SUPPORTED' => HostCardEmulationStatus.functionNotSupported,
    'TAG_DISCONNECTED' => HostCardEmulationStatus.tagDisconnected,
    _ =>
      throw ArgumentError.value(
        status,
        'status',
        'No enum value with that status.',
      ),
  };
}
